File: 백준/Gold/24042. 횡단보도/횡단보도.py
import heapq
import sys

def dijkstra():
    q = []
    # cost,cur
    heapq.heappush(q,(0,1))
    dist = [sys.maxsize]*(n+1)
    dist[1] = 0
    while q:
        cost,cur = heapq.heappop(q)
        if cur==n:
            return cost
        if dist[cur] < cost:
            continue
        for next,i in arr[cur]:
            next_cost = ((cost-i+m-1)//m)*m+i+1
            if dist[next]>next_cost:
                dist[next] = next_cost
                heapq.heappush(q,(next_cost,next))

n,m = map(int,input().split())
arr = [[] for _ in range(n+1)]
for i in range(m):
    a,b = map(int,input().split())
    arr[a].append((b,i))
    arr[b].append((a,i))

answer = dijkstra()
print(answer)

File: 백준/Gold/24042. 횡단보도/횡단보도.java
import java.util.*;
import java.lang.*;
import java.io.*;

class Main {
    static private int N,M;
    static private List<int[]>[] arr;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        arr = new ArrayList[N+1];
        for(int i=0;i<=N;i++){
            arr[i] = new ArrayList<>();
        }
        for(int i=0;i<M;i++){
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            arr[a].add(new int[]{b,i});
            arr[b].add(new int[]{a,i});
        }
        long[] dist = new long[N+1];
        Arrays.fill(dist,Long.MAX_VALUE);
        dist[1] = 0;
        //cost,cur
        PriorityQueue<Node> pq = new PriorityQueue<>((o1,o2)->{
            return Long.compare(o1.cost,o2.cost);
        });
        pq.add(new Node(0L,1));
        while(!pq.isEmpty()){
            Node cur = pq.poll();
            if(cur.idx==N){
                System.out.println(cur.cost);
                return;
            }
            if(dist[cur.idx]<cur.cost){
                continue;
            }
            for(int[] next : arr[cur.idx]){
                long nextCost = ((cur.cost-next[1]+M-1)/M)*M+next[1]+1;
                if(dist[next[0]] > nextCost){
                    dist[next[0]] = nextCost;
                    pq.add(new Node(nextCost, next[0]));
                }
            }
        }
    }

    static class Node{
        Long cost;
        int idx;
        public Node(Long cost, int idx){
            this.cost = cost;
            this.idx = idx;
        }
    }
}